### PR TITLE
Use `EpochNo` instead of `SlotNo` in `Pool`, `Cert` and `Certs` rules

### LINKED ISSUE #4692: Pass `curEpochNo` instead of `curSlotNo` to all the rules that don't need `SlotNo`
Rules like `POOL` do not really care about the current slot number, but instead they convert it into the current epoch number. In order to make it clearer and more efficient it would make more sense to convert the slot number to an epoch number in a top level rule like `LEDGERS` and pass it down through the environment.

### PR DESCRIPTION
**Pull request by teodanciu**
# Description

Slightly simplify rules by not passing slotNo around when not required.

Closes https://github.com/IntersectMBO/cardano-ledger/issues/4692

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

### CODE REVIEW

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Delpl.hs [hunk 1]
```diff
@@ -196,14 +197,18 @@ delplTransition ::
...
+  , Era era
   ) =>
   TransitionRule (ShelleyDELPL era)
 delplTransition = do
   TRC (DelplEnv slot ptr pp acnt, d, c) <- judgmentContext
   case c of
     ShelleyTxCertPool poolCert -> do
+      cEpoch <- liftSTS $ do
+        ei <- asks epochInfoPure
+        epochInfoEpoch ei slot
```
Comment (lehins) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Delpl.hs:209]: The ultimate goal of this PR should be to make this call done only once! That is why description of the ticket says it should be done somewhere in `LEDGERS` rule.

The point is that we do not want to compute the same value over and over again for every transaction in a block, and here even worse for every certificate in a transaction. (Note that `checkSlotNotTooLate` also does the same computation with `epochInfoEpoch`) In other words we need to reduce number of calls to logic in `epochInfoEpoch` to an absolute minimum.

It does mean that some rules (like `DELPL`) will have to accept both: current slot and current epoch. That is a small price to pay in complexity for avoiding the same calculation done a whole bunch of times during validation of every block.

**Review by lehins — CHANGES_REQUESTED** [review]
@teodanciu I think you misread the ticket or maybe it just doesn't explain the goal clearly.

The goal is to reduce computation. Se my comment in the PR

**Comment by lehins** [comment]
FYI. `applyTx` from `Cardano.Ledger.Shelley.API.Mempool` will also be affected by my comment. Which has two alternative solutions that I can think of, but we can talk about if you run into problems in the `ApplyTx` interface

**Comment by teodanciu** [comment]
Closing, since i redid all in this PR: https://github.com/IntersectMBO/cardano-ledger/pull/4765

**Pull request closed without merging**